code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/model/SubmissionsGraph.kt

```diff
@@ -36,26 +36,35 @@ class GraphTransformer<T>(
     private val submissionsGraphContext: SubmissionsGraphContext<T>,
     private val graph: SubmissionsGraphAlias
 ) {
-    private val vertexByCode = HashMap<String, SubmissionsNode>()
+    private val vertexByUnifiedCode = HashMap<String, SubmissionsNode>()
+    private val vertexByInitialCode = HashMap<String, SubmissionsNode>()
     private val idFactory = IdentifierFactoryImpl()
 
     init {
         graph.vertexSet().forEach {
-            vertexByCode[it.code] = it
+            vertexByInitialCode[it.code] = it
         }
     }
 
     fun add(submission: Submission) {
```

Comment:
Also, I think we can rewrite this function since it has the same pieces of code:
```kotlin
    fun add(submission: Submission) {
        submissionsGraphContext.unifier.run {
            vertexByInitialCode.compute(submission) {
                val unifiedSubmission = submission.unify()
                vertexByUnifiedCode.compute(unifiedSubmission) {
                    SubmissionsNode(unifiedSubmission, idFactory.uniqueIdentifier()).also {
                        graph.addVertex(it)
                    }
                }
            }
        }
    }

    private fun HashMap<String, SubmissionsNode>.compute(submission: Submission, newNodeAction: (Submission) -> SubmissionsNode?) =
        this.compute(submission.code) { _, vertex ->
            vertex?.also {
                // Update existing vertex
                vertex.submissionsList.add(submission.info)
            } ?: newNodeAction(submission)
        }
```

What do you think?